kep/260-network-aware-scheduling/README.md

```diff
@@ -0,0 +1,1398 @@
+    - [Scalability](#scalability)
+    - [Troubleshooting](#troubleshooting)
+- [Graduation criteria](#graduation-criteria)
+- [Implementation history](#implementation-history)
+<!-- /toc -->
+
+# Summary
+
+This proposal describes the behavior of the Network-Aware Scheduling framework
+that considers latency and bandwidth in the scheduling decision-making process.
+
+The proposal introduces two custom resources, **AppGroup CRD** and **NetworkTopology CRD**, to maintain 
+both the service topology information from application microservice dependencies and the infrastructure 
+network topology where network weights are established between regions and zones in the cluster. 
+Thus, both application and infrastructure network topology are considered during scheduling. 
+
+The proposal also presents a bandwidth resource component (DaemonSet) to advertise bandwidth resources 
+as **extended resources** to allow already available filter/scoring plugins (e.g., `PodFitsResources`, 
+`BalancedAlocation`) to consider bandwidth allocation.
+
+To tackle latency and bandwidth in the scheduling process, three plugins are proposed in this document: 
+- `TopologicalSort` (**QueueSort**).
+- `NodeNetworkCostFit` (**Filter**).
+- `NetworkMinCost` (**Score**).
+
+# Motivation
+
+Many applications are latency-sensitive, demanding lower latency between microservices in the application.
+Scheduling policies that aim to reduce costs or increase resource efficiency are not enough for applications
+where end-to-end latency becomes a primary objective.
+Applications such as the Internet of Things (IoT), multi-tier web services, and video streaming services
+would benefit the most from network-aware scheduling policies, which consider latency and bandwidth
+in addition to the default resources (e.g., CPU and memory) used by the scheduler.
+
+Users encounter latency issues frequently when using multi-tier applications.
+These applications usually include tens to hundreds of microservices with complex interdependencies/chains.
+Distance from servers with chained microservices deployed is usually the primary culprit.
+The best strategy is to reduce the latency between chained microservices in the same application,
+according to the prior work about [Service Function Chaining](https://www.sciencedirect.com/science/article/pii/S1084804516301989) (SFC).
+
+Besides, bandwidth plays an essential role for those applications with high volumes of data transfers among microservices.
+For example, multiple replicas in a database application may require frequent copies to ensure data consistency.
+[Spark jobs](https://spark.apache.org/) may have frequent data transfers among map and reduce nodes.
+Insufficient network capacity in nodes would lead to increasing delay or packet drops,
+which will degrade the quality of service for applications.
+
+We propose a **Network-Aware Scheduling framework** for Kubernetes that focus on delivering low latency to end-users
```

Comment:
@Huang-Wei , in network aware scheduling,  the network latency and bandwidth have to be considered on connections between nodes and dependency requirements between microservices. That's why we think both the microservice interdependencies and the physical network topology are needed in this KEP.

Yes, I agree that the DAG-like requirement is another bigger scope. However, if we only consider the fitting of one pod's inbound and outbound bandwidth requirements on one node's bandwidth capacity and ignore the actual communications between pods and nodes, it would be less applicable/useful to real-world applications. Implementation-wise, we do not need to add any plugins to achieve that; we only need to patch the node with network bandwidth as extended resources.

Considering that the per-node topology-aware scheduling has been proposed and the network topology-aware scheduling is still missing, we want to add this DAG awareness to make network-aware scheduler plugins usable in most use cases, including HPC workloads.

Sorry for adding more workloads for reviewers. The latency-aware plugin we already presented before was DAG-aware. The additional DAG awareness added is on bandwidth allocations between nodes.  For nodes within a zone, we believe the network capacity is only bounded by inbound and outbound capacity on each node,  so it would be the same as per node allocation. However, we do not want to exclude the use case where pods are placed on nodes in different zones. In that case, the available bandwidth is really bounded by connections between nodes. That's why our `maxNetworkCost` and `minBandwidth` requirements are listed under the dependencies but not as extended resources on pods.

If you believe it is needed, we can present the DAG awareness part again in the community meeting. We think it is not sufficient to consider only bandwidth allocation on nodes for network-aware scheduling because bandwidth allocation is really on connections between nodes. 


Besides, even though we add DAG awareness in this KEP, I think there is still a more significant scope missing about the topology-aware scheduling, as you mentioned.  There might be a need to unify all physical topology descriptions into one, including node topology and network topology. And we may also want to unify all kinds of logical topology descriptions for applications.